zcash_primitives/src/transaction/txid.rs

```diff
@@ -0,0 +1,580 @@
+    let mut nh = HashWriter::new(ZCASH_SAPLING_OUTPUTS_NONCOMPACT_HASH_PERSONALIZATION);
+    for s_out in shielded_outputs {
+        ch.write_all(&s_out.cmu.to_repr().as_ref()).unwrap();
+        ch.write_all(&s_out.ephemeral_key.to_bytes()).unwrap();
+        ch.write_all(&s_out.enc_ciphertext[..52]).unwrap();
+
+        mh.write_all(&s_out.enc_ciphertext[52..564]).unwrap();
+
+        nh.write_all(&s_out.cv.to_bytes()).unwrap();
+        nh.write_all(&s_out.enc_ciphertext[564..]).unwrap();
+        nh.write_all(&s_out.out_ciphertext).unwrap();
+    }
+
+    let mut h = HashWriter::new(ZCASH_SAPLING_OUTPUTS_HASH_PERSONALIZATION);
+    h.write_all(&ch.finalize().as_bytes()).unwrap();
+    h.write_all(&mh.finalize().as_bytes()).unwrap();
+    h.write_all(&nh.finalize().as_bytes()).unwrap();
+    h.finalize()
+}
+
+/// The txid commits to the hash of all transparent outputs. The
+/// prevout and sequence_hash components of txid
+fn transparent_digests<A: transparent::Authorization>(
+    bundle: &transparent::Bundle<A>,
+) -> TransparentDigests<Blake2bHash> {
+    TransparentDigests {
+        prevout_digest: transparent_prevout_hash(&bundle.vin),
+        sequence_digest: transparent_sequence_hash(&bundle.vin),
+        outputs_digest: transparent_outputs_hash(&bundle.vout),
+        per_input_digest: None,
+    }
+}
+
+#[cfg(feature = "zfuture")]
+fn tze_digests<A: tze::Authorization>(bundle: &tze::Bundle<A>) -> TzeDigests<Blake2bHash> {
+    // The txid commits to the hash for all outputs.
+    TzeDigests {
+        inputs_digest: hash_tze_inputs(&bundle.vin),
+        outputs_digest: hash_tze_outputs(&bundle.vout),
+        per_input_digest: None,
+    }
+}
+
+fn hash_header_txid_data(
+    version: TxVersion,
+    // we commit to the consensus branch ID with the header
+    consensus_branch_id: BranchId,
+    lock_time: u32,
+    expiry_height: BlockHeight,
+) -> Blake2bHash {
+    let mut h = HashWriter::new(ZCASH_HEADERS_HASH_PERSONALIZATION);
+
+    (&mut h)
+        .write_u32::<LittleEndian>(version.header())
+        .unwrap();
+    (&mut h)
+        .write_u32::<LittleEndian>(version.version_group_id())
+        .unwrap();
+    (&mut h)
+        .write_u32::<LittleEndian>(consensus_branch_id.into())
+        .unwrap();
+    (&mut h).write_u32::<LittleEndian>(lock_time).unwrap();
+    (&mut h)
+        .write_u32::<LittleEndian>(expiry_height.into())
+        .unwrap();
+
+    h.finalize()
+}
+
+fn hash_transparent_txid_data(t_digests: Option<&TransparentDigests<Blake2bHash>>) -> Blake2bHash {
+    let mut h = HashWriter::new(ZCASH_TRANSPARENT_HASH_PERSONALIZATION);
+    if let Some(d) = t_digests {
+        h.write_all(d.prevout_digest.as_bytes()).unwrap();
+        h.write_all(d.sequence_digest.as_bytes()).unwrap();
+        h.write_all(d.outputs_digest.as_bytes()).unwrap();
+        if let Some(s) = d.per_input_digest {
+            h.write_all(s.as_bytes()).unwrap();
+        };
```

Comment:
Should `per_input_digest` be included in the `transparent_digest`? https://zips.z.cash/zip-0244#t-2-transparent-digest